File: src/main/java/com/cjcrafter/openai/chat/ChatMessage.java
package com.cjcrafter.openai.chat;

import com.google.gson.JsonObject;

/**
 * The ChatGPT API takes a list of 'roles' and 'content'. The role is
 * one of 3 options: system, assistant, and user. 'System' is used to
 * prompt ChatGPT before the user gives input. 'Assistant' is a message
 * from ChatGPT. 'User' is a message from the human.
 */
public final class ChatMessage {

    private final String role;
    private final String content;

    /**
     * Constructor requires who sent the message, and the content of the
     * message.
     *
     * @param role    Who sent the message.
     * @param content The raw content of the message.
     */
    public ChatMessage(String role, String content) {
        this.role = role;
        this.content = content;
    }

    public ChatMessage(JsonObject json) {
        this(json.get("role").getAsString(), json.get("content").getAsString());
    }

    public String getRole() {
        return this.role;
    }

    public String getContent() {
        return this.content;
    }
}

File: src/main/kotlin/com/cjcrafter/openai/chat/ChatMessage.kt
package com.cjcrafter.openai.chat

import com.cjcrafter.openai.chat.tool.ToolCall
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

/**
 * ChatGPT's biggest innovation is its conversation memory. To remember the
 * conversation, we need to map each message to who sent it. This data class
 * wraps a message with the user who sent the message.
 *
 * @property role The user who sent this message.
 * @property content The string content of the message.
 * @see ChatUser
 */
data class ChatMessage @JvmOverloads constructor(
    var role: ChatUser,
    @JsonInclude var content: String?, // JsonInclude here is important for tools
    @JsonProperty("tool_calls") var toolCalls: List<ToolCall>? = null,
    @JsonProperty("tool_call_id") var toolCallId: String? = null,
) {
    init {
        if (role == ChatUser.TOOL) {
            requireNotNull(toolCallId) { "toolCallId must be set when role is TOOL" }
        }
    }

    companion object {

        /**
         * Returns a new [ChatMessage] using [ChatUser.SYSTEM].
         */
        @JvmStatic
        fun String.toSystemMessage(): ChatMessage {
            return ChatMessage(ChatUser.SYSTEM, this)
        }

        /**
         * Returns a new [ChatMessage] using [ChatUser.USER].
         */
        @JvmStatic
        fun String.toUserMessage(): ChatMessage {
            return ChatMessage(ChatUser.USER, this)
        }

        /**
         * Returns a new [ChatMessage] using [ChatUser.ASSISTANT].
         */
        @JvmStatic
        fun String.toAssistantMessage(): ChatMessage {
            return ChatMessage(ChatUser.ASSISTANT, this)
        }

        /**
         * Returns a new [ChatMessage] using [ChatUser.TOOL].
         */
        @JvmStatic
        fun String.toToolMessage(toolCallId: String): ChatMessage {
            return ChatMessage(ChatUser.TOOL, this, toolCallId = toolCallId)
        }
    }
}